music/src/ddsp/pitch_utils.ts

```diff
@@ -0,0 +1,191 @@
+    SPICE_MODEL_MULTIPLE; //always multiple of 512;
+  const fullInputWithPadding = inputTensor.pad([
+    [0, inputSampleNum - audioChannelDataLength],
+  ]);
+  const expectedDuration = fullInputWithPadding.size / SPICE_SAMPLE_RATE;
+
+  // determine if slicing is needed
+  const output = await spiceModel.execute({
+    input_audio_samples: fullInputWithPadding,
+  });
+  // @ts-ignore
+  let uncertainties = await output[0].data();
+  // @ts-ignore
+  const pitches = await output[1].data();
+
+  //spice extracts 1 pitch for every 32ms, so to get duration,
+  // we multiply it by 32
+  //we minus one to offset for the zero
+  //to convert it into seconds, we divide by 1000
+  if (((pitches.length - 1) * 32) / 1000 === expectedDuration) {
+    let lastPitch = 20.0;
+
+    for (let i = 0; i < pitches.length; ++i) {
+      const confidence = 1.0 - uncertainties[i];
+      allConfidences.push(confidence);
+      if (confidence >= CONF_THRESHOLD) {
+        lastPitch = getPitchHz(pitches[i]);
+        spicePitchesOutput.push(lastPitch);
+      } else {
+        const noise = tf.truncatedNormal([1], 0.0, PITCH_CONF_JITTER);
+
+        // @ts-ignore
+        const jitter = 1.0 - noise.dataSync();
+        spicePitchesOutput.push(lastPitch * jitter);
+        noise.dispose();
+      }
+    }
+  } else {
+    // we try splicing it into 4 parts to run spice again, then combine it all
+    const finalPitchesLength = inputSampleNum / SPICE_MODEL_MULTIPLE + 1;
+    const stitchedPitches = new Float32Array(finalPitchesLength);
+    uncertainties = new Float32Array(finalPitchesLength);
+    for (let i = 0; i < inputSampleNum; i += inputSampleNum / 4) {
+      const partialInput = fullInputWithPadding.slice(
+        [i],
+        [inputSampleNum / 4]
+      );
+      const partialOutput = await spiceModel.execute({
+        input_audio_samples: partialInput,
+      });
+      // @ts-ignore
+      const partialUncertainties = await partialOutput[0].data();
+      // @ts-ignore
+      const partialPitches = await partialOutput[1].data();
+      const index = Math.floor(i / SPICE_MODEL_MULTIPLE);
+      uncertainties.set(partialUncertainties, index);
+      stitchedPitches.set(partialPitches, index);
+      partialInput.dispose();
+      // @ts-ignore
+      partialOutput[0].dispose();
+      // @ts-ignore
+      partialOutput[1].dispose();
+    }
+
+    let lastPitch = 20.0;
+
+    for (let i = 0; i < stitchedPitches.length; ++i) {
+      const confidence = 1.0 - uncertainties[i];
+      allConfidences.push(confidence);
+      if (confidence >= CONF_THRESHOLD) {
+        lastPitch = getPitchHz(stitchedPitches[i]);
+        spicePitchesOutput.push(lastPitch);
+      } else {
+        const noise = tf.truncatedNormal([1], 0.0, PITCH_CONF_JITTER);
+
+        // @ts-ignore
+        const jitter = 1.0 - noise.dataSync();
+        spicePitchesOutput.push(lastPitch * jitter);
+        noise.dispose();
+      }
+    }
+  }
+
+  // @ts-ignore
+  output[0].dispose();
```

Comment:
Why are there so many ts-ignores? I don't think normally we need to ignore the tensor disposing

Reply:
@notwaldorf I have a TODO to fix types (which I privately DM'ed Jesse) so apologies for not stating it in this draft PR! (It's also why this is still a draft PR and not ready for a review)

In this case, I'm actually struggling a little with TFJS and TS where they expect a Tensor in the return of a prediction but our models actually return an array with 2 tensors.